crates/ignore/src/types.rs

```diff
@@ -112,35 +114,241 @@ use crate::{Error, Match};
+
+type FileTypeIndicesBitArray = BitArr!(for FILE_TYPE_INDICES_BIT_ARRAY_LEN);
+
+fn bit_array_into_file_type_indices_iterator(
+    bit_array: FileTypeIndicesBitArray,
+) -> FileTypeIndicesBitArrayIterator {
+    FileTypeIndicesBitArrayIterator::new(bit_array)
+}
+
+pub struct FileTypeIndicesBitArrayIterator {
+    bit_array_iterator:
+        Enumerate<<FileTypeIndicesBitArray as IntoIterator>::IntoIter>,
+}
+
+impl FileTypeIndicesBitArrayIterator {
+    pub fn new(bit_array: FileTypeIndicesBitArray) -> Self {
+        Self { bit_array_iterator: bit_array.into_iter().enumerate() }
+    }
+}
+
+impl Iterator for FileTypeIndicesBitArrayIterator {
+    type Item = usize;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some((file_type_index, is_selected)) =
+            self.bit_array_iterator.next()
+        {
+            if is_selected {
+                return Some(file_type_index);
+            }
+        }
+        None
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum FileTypeIndicesToken {
+    Empty,
+    Single(usize),
+    BitArray(FileTypeIndicesBitArray),
+    Vec(Vec<usize>),
+}
+
+impl FileTypeIndicesToken {
+    pub fn push(&mut self, file_type_index: usize) {
+        if file_type_index < FILE_TYPE_INDICES_BIT_ARRAY_LEN {
+            match self {
+                Self::Empty => {
+                    *self = Self::Single(file_type_index);
+                }
+                Self::Single(first_file_type_index) => {
+                    let mut bit_array = BitArray::ZERO;
+                    bit_array.set(*first_file_type_index, true);
+                    bit_array.set(file_type_index, true);
+                    *self = Self::BitArray(bit_array);
+                }
+                Self::BitArray(bit_array) => {
+                    bit_array.set(file_type_index, true);
+                }
+                Self::Vec(vec) => {
+                    vec.push(file_type_index);
+                }
+            }
+        } else {
+            match self {
+                Self::Empty => {
+                    *self = Self::Vec(vec![file_type_index]);
+                }
+                Self::Single(first_file_type_index) => {
+                    *self = Self::Vec(vec![
+                        *first_file_type_index,
+                        file_type_index,
+                    ]);
+                }
+                Self::BitArray(bit_array) => {
+                    let mut vec: Vec<_> =
+                        bit_array_into_file_type_indices_iterator(*bit_array)
+                            .collect();
+                    vec.push(file_type_index);
+                    *self = Self::Vec(vec);
+                }
+                Self::Vec(vec) => {
+                    vec.push(file_type_index);
+                }
+            }
+        }
+    }
+}
+
+impl Default for FileTypeIndicesToken {
+    fn default() -> Self {
+        Self::Empty
+    }
+}
+
+impl IntoIterator for FileTypeIndicesToken {
```

Comment:
I implemented `IntoIterator` for the owned and reference versions of this type, at different times it looked like I'd be using the one or the other